Be verbose about failure to import airflow_local_settings

baryluk opened the pull request:
Currently, if the module exists, but has errors (for example syntax
error, or transitive import of module that does not exist),
the airflow scheduler will not show any error.

In my opinion `ImportError` of `airflow_local_settings` should
be a fatal error, but let it be a warning for now.

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

baryluk committed:
Be verbose about failure to import airflow_local_settings

Currently, if the module exists, but has errors (for example syntax
error, or transitive import of module that does not exist),
the airflow scheduler will not show any error.

In my opinion `ImportError` of `airflow_local_settings` should
be a fatal error, but let it be a warning for now.

baryluk committed:
Do not ignore transitive import failures from airflow_local_settings

A case of `airflow_local_settings.py` importing a module
that does not exist, will also throw `ModuleNotFoundError`,
but it should not be silently ignored.

potiuk commented on airflow/settings.py:
```diff
@@ -432,7 +432,12 @@ def import_local_settings():
     except ImportError:
-        log.debug("Failed to import airflow_local_settings.", exc_info=True)
+        log.warning("Failed to import airflow_local_settings.", exc_info=True)
+    except ModuleNotFoundError as e:
```
ModuleNotFoundError derirves from ImportError so it should be higher.

baryluk replied:
Done. Also made it critical and re-raise the exception.

This is no different that other possible errors, like `SyntaxError`.

baryluk committed:
Reorder ModuleNotFoundError, make it critical

`ModuleNotFoundError` is a subclass of `ImportError`, so must be handled first.

potiuk commented on airflow/settings.py:
```diff
@@ -431,8 +431,15 @@ def import_local_settings():
+            log.debug("No airflow_local_settings to import.", exc_info=True)
+        else:
+            log.critical("Failed to import airflow_local_settings due to a transitive module not found error.", exc_info=True)
+            raise e
```
```suggestion
            raise
```

baryluk replied:
Thanks, that is indeed better. I was not aware of that feature of Python.

potiuk commented on airflow/settings.py:
```diff
@@ -431,8 +431,15 @@ def import_local_settings():
+            raise e
+    except ImportError as e:
+        log.critical("Failed to import airflow_local_settings.", exc_info=True)
+        raise e
```
```suggestion
        raise
```

potiuk reviewed (changes requested):
Agree we should raise error but `raise` is better as it keeps location of the original exception.

baryluk committed:
Use raise to reraise

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

baryluk committed:
Use raise to reraise

Co-authored-by: Jarek Potiuk <jarek@potiuk.com>

potiuk reviewed (approved):
Nice! Really useful to detect some hard-to-debug problems!

baryluk committed:
Remove unused variable name in except

baryluk committed:
Reformat long function invocation

Makes black happy

baryluk committed:
Update test_settings.py
